awscli/alias.py

```diff
@@ -0,0 +1,206 @@
+            retrieved from `AliasLoader.get_aliases()[alias_name]`
+
+        :type session: botocore.session.Session
+        :param session: The botocore session
+
+        :type command_table: dict
+        :param command_table: The command table containing all of the
+            possible service command objects that a particular alias could
+            redirect to.
+
+        :type parser: awscli.argparser.MainArgParser
+        :param parser: The parser to parse commands provided at the top level
+            of a CLI command which includes service commands and global
+            parameters. This is used to parse the service commmand and any
+            global parameters from the alias's value.
+        """
+        super(ServiceAliasCommand, self).__init__(alias_name, alias_value)
+        self._session = session
+        self._command_table = command_table
+        self._parser = parser
+
+    def __call__(self, args, parsed_globals):
+        alias_args = self._get_alias_args()
+        parsed_alias_args, remaining = self._parser.parse_known_args(
+            alias_args)
+        self._update_parsed_globals(parsed_alias_args, parsed_globals)
+        # Take any of the remaining arguments that were not parsed out and
+        # prepend them to the remaining args provided to the alias.
+        remaining.extend(args)
+        LOG.debug(
+            'Alias %r passing on arguments: %r to %r command',
+            self._alias_name, remaining, parsed_alias_args.command)
+        # Pass the update remaing args and global args to the service command
+        # the alias proxied to.
+        return self._command_table[parsed_alias_args.command](
+            remaining, parsed_globals)
+
+    def _get_alias_args(self):
+        try:
+            alias_args = shlex.split(self._alias_value)
+        except ValueError as e:
+            raise InvalidAliasException(
+                'Value of alias "%s" could not be parsed. '
+                'Received error: %s when parsing:\n%s' % (
+                    self._alias_name, e, self._alias_value)
+            )
+
+        alias_args = [arg.strip(os.linesep) for arg in alias_args]
+        LOG.debug(
+            'Expanded subcommand alias %r with value: %r to: %r',
+            self._alias_name, self._alias_value, alias_args
+        )
+        return alias_args
+
+    def _update_parsed_globals(self, parsed_alias_args, parsed_globals):
+        global_params_to_update = self._get_global_parameters_to_update(
+            parsed_alias_args)
+        # Emit the top level args parsed event to ensure all possible
+        # customizations that typically get applied are applied to the
+        # global parameters provided in the alias before updating
+        # the original provided global parameter values
+        # and passing those onto subsequent commands.
+        emit_top_level_args_parsed_event(self._session, parsed_alias_args)
```

Comment:
I decided to go with the approach of re-emitting the top level args event because it is really the only reliable way to recreate the final value of parsed globals as many of them contain customizations that modify the value.